File: android/src/main/java/expo/modules/splashfullscreen/SplashScreenModule.kt
package expo.modules.splashfullscreen

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.lang.ref.WeakReference

class SplashScreenModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("SplashScreenModule")

    Events("didShow", "didHide", "didFail")

    OnCreate {
      // Hold the module weakly so an RN reload that destroys this module doesn't leak via the
      // closure stored on the overlay singleton (which is app-scoped).
      val weakSelf = WeakReference(this@SplashScreenModule)
      SplashScreenOverlay.setEventEmitter { name, body ->
        weakSelf.get()?.sendEvent(name, body)
      }
    }

    OnDestroy {
      SplashScreenOverlay.setEventEmitter(null)
    }

    AsyncFunction("showFullScreen") {
      SplashScreenOverlay.showFullScreen()
    }

    AsyncFunction("hide") { options: Map<String, Any?> ->
      val fade = options["fade"] as? Boolean ?: true
      val duration = when (val d = options["duration"]) {
        is Number -> d.toLong()
        else -> 300L
      }
      SplashScreenOverlay.hide(fade, duration)
    }
  }
}

File: ios/SplashScreenModule.swift
import ExpoModulesCore

public class SplashScreenModule: Module {
  public func definition() -> ModuleDefinition {
    Name("SplashScreenModule")

    Events("didShow", "didHide", "didFail")

    OnCreate {
      // Bridge module's sendEvent into the overlay singleton, weak so a JS reload that destroys
      // this module doesn't keep it alive via the closure stored on the singleton.
      SplashScreenOverlay.shared.setEventEmitter { [weak self] name, body in
        self?.sendEvent(name, body)
      }
    }

    OnDestroy {
      SplashScreenOverlay.shared.setEventEmitter(nil)
    }

    AsyncFunction("showFullScreen") {
      SplashScreenOverlay.shared.showFullScreen()
    }

    AsyncFunction("hide") { (options: [String: Any]) in
      let fade = (options["fade"] as? Bool) ?? true
      let duration = (options["duration"] as? Double) ?? 300
      // Plugin layer guarantees ms; convert to seconds for UIView.animate.
      SplashScreenOverlay.shared.hide(fade: fade, duration: duration / 1000)
    }
  }
}
